### PR TITLE
Implement pseudo binary search algorithm for finding the first visible starting index

### PR DESCRIPTION
**Pull request by tim-evans**
This needs some review, since this stuff makes my head hurt.

I think it works properly.

### CODE REVIEW

**Comment by tim-evans** [comment]
The actual block of code in a more readable format is here:

https://github.com/stefanpenner/pack/blob/fcbaba5fbd2603d03241edd4645298cefeac7b14/index.js#L263-L317

**Comment by stefanpenner** [comment]
oooooo, I will review thoroughly Thursday. I am excite! (it will be my list-view day :))

File: index.js [hunk 1]
```diff
@@ -257,31 +255,62 @@ ShelfFirst.prototype.visibleStartingIndex = function(topOffset, width) {
...
-  var position, entry;
-  var previousTop = 0;
-  var index = -1;
+  var height = this.height();
   var length = this.length();
 
-  // TODO: implement fuzzy binary search for large N to reduce cost when search
-  // space is large. Should get this closer to O(log n)
+  // Start searching using the last item in the list
+  // and the bottom of the list for calculating the average hegiht
```
Comment (stefanpenner) [inline index.js]: typo hegiht -> height

Comment (stefanpenner) [inline index.js]: should likely leave the `O(log n)` goal (or something more accurate) statement so future travelers understand the motivation for the bellow algorithim

File: test.js [hunk 2]
```diff
@@ -97,7 +97,7 @@ assert.equal(bin.visibleStartingIndex(175, 100), 7);
 assert.equal(bin.visibleStartingIndex( 49, 100), 0);
 
 assert.equal(bin.visibleStartingIndex(200, 100), 7);
-assert.equal(bin.visibleStartingIndex(400, 100), 0); // TODO: eventually this should be 7
+assert.equal(bin.visibleStartingIndex(400, 100), 7);
```
Comment (stefanpenner) [inline test.js:100]: woohoo!

**Comment by tim-evans** [comment]
Awesome!

By the way, this doesn't handle ragged rows. (I'll try to make an illustration of this)

**Comment by stefanpenner** [comment]
> By the way, this doesn't handle ragged rows. (I'll try to make an illustration of this)

got time for that illustration?

sorry I let this slip, mind helping with a quick rebase? I'll get it merged in shortly after that.

**Comment by tim-evans** [comment]
artboard 1

**Comment by tim-evans** [comment]
This is a good example of how https://www.artsy.net/shows looks (without text), and illustrates ragged rows fairly well. (In the last row, the center image is the one farthest to the bottom; the top row has the first image as the farthest one to the bottom)

**Comment by stefanpenner** [comment]
@tim-evans ah correct, we don't currently support this. Although i would love to at some point.

**Comment by tim-evans** [comment]
Goodie! Then my algorithm is ok, since it's not supported :)

**Comment by stefanpenner** [comment]
@tim-evans i suspect we will need to add additional packing algorithims, that support various masonry styles. Although the memoizing variants we need are tricky.

**Comment by tim-evans** [comment]
Agreed. It's tricky stuff.

**Pull request merged**